### PR TITLE
ISPN-11896 non blocking cache store SPI docs

### PR DESCRIPTION
**Pull request by oraNod**
https://issues.redhat.com/browse/ISPN-11896

### CODE REVIEW

**Comment by oraNod** [comment]
@wburns hi. these are all just suggestions for phrasing along with stuff like grammar + punctuation. feel free to leave or take anything. I'd also be happy to go through anything here and explain some of the suggested changes.

there are places where I've changed passive voice to active voice. also changed tense from future to present, which can help improve clarity. but I've been careful to avoid changing meaning, so hopefully there's nothing here that loses technical accuracy.

File: core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java [hunk 1]
```diff
@@ -87,146 +83,145 @@
...
-    * will not be invoked on this store until after the returned Stage completes.
+    * This method is guaranteed not to be invoked concurrently with other operations. This means other methods are
+    * not invoked on this store until after the returned Stage completes.
     * <p>
     * It is expected that an implementation should be able to "restart" by invoking {@code start} a second time if
     * {@link #stop()} has been invoked and allowed for its stage to complete.
-    * @param ctx initialization context used to initialize this store
-    * @return a stage that when complete signals that this store has been successfully started
+    * @param ctx initialization context used to initialize this store.
+    * @return a stage that, when complete, signals that this store has started successfully.
```
Comment (oraNod) [inline core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java:176]: @wburns here do we lose meaning if we change "signals" to "indicates"? or is "signals" terminology that is consistent with rxjava etc?

File: core/src/main/java/org/infinispan/persistence/spi/InitializationContext.java [hunk 2]
```diff
@@ -51,17 +51,18 @@
...
 
    /**
-    * Returns an executor that Infinispan uses internally for non blocking tasks. The user must guarantee tasks
-    * submitted to this executor will not block the thread it is ran on. Failure to do so can slow down Infinispan's
-    * handling of operations as these threads are limited to the number of cores and are used extensively.
-    * @return an executor that can be used to submit tasks that will not block the thread it runs on
+    * Returns an executor that Infinispan uses to perform tasks in a non-blocking manner. Users must guarantee that the tasks they
+    * submit to this executor do not block the thread in which the executor runs. Doing so can cause Infinispan to handle operations
+    * more slowly, reducing performance, because threads are limited to the number of cores and are used extensively.
+    * @return an executor that can submit non-blocking tasks.
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/persistence/spi/InitializationContext.java:57]: How about just "an executor for non-blocking tasks"?

File: core/src/main/java/org/infinispan/persistence/spi/InitializationContext.java [hunk 3]
```diff
@@ -51,17 +51,18 @@
...
    Executor getNonBlockingExecutor();
 
    /**
-    * Returns a manager that is designed to execute tasks that may block. This manager ensures that only the blocking
-    * portion is ran on a blocking thread and any stage is continued on a non blocking thread.
-    * @return a manager that should be used to execute blocking operations
+    * Returns a manager that can execute thread-blocking operations. The manager runs any blocking operations on blocking threads,
+    * any stages continue on non-blocking threads. Use this manager to ensure that blocking operations do not leak into the
+    * Infinispan internal system.
+    * @return a manager that can execute blocking operations.
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/persistence/spi/InitializationContext.java:65]: IMO the original better conveys the purpose of the `BlockingManager`.

Comment (oraNod) [reply core/src/main/java/org/infinispan/persistence/spi/InitializationContext.java]: thanks @ryanemerson I've changed back to the original (with a couple minor tweaks)

File: core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java (hunk 1 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java:176]: No, signal is just a term I used here. This is specific to CompletionStage as well at this point, so I am fine either way.

File: core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java [hunk 4]
```diff
@@ -24,53 +24,49 @@
...
- * {@link StoreConfiguration#segmented()} available from the {@code InitializationContext}.
+ * Although recommended, segmentation support in store implementations is optional. Segment parameters are provided
+ * for all methods where segment information is required, for example {@link #load(int, Object)} and
+ * {@link #publishEntries(IntSet, Predicate, boolean). If your store implementation does not support segmentation,
+ * you can ignore these parameters. However, you should note that segmented stores allow Infinispan caches to more
+ * efficiently perform bulk operations such as {@code Cache.size()} or {@code Cache.entrySet().stream()}. Segmentation
+ * also decreases the duration of state transfers when {@link PersistenceConfiguration#fetchPersistentState()} is enabled,
+ * as well as the time required to remove data by segments. To indicate that a store implementation supports segmentation,
+ * the {@link Characteristic#SEGMENTABLE} characteristic must be returned by the {@link #characteristics()} method. Store
+ * implementations can determine if stores are segmented if {@link StoreConfiguration#segmented()} is enabled, which is
```
[the remark this answers is not in the document]
Comment (oraNod) [reply core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java]: @wburns hmm. thanks, I updated but maybe the phrasing still isn't 100%. could you look at this line again please?

File: core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java [hunk 5]
```diff
@@ -87,146 +83,145 @@
...
     possible for the cache operations to be accepted in the interim period between the loss of availability and the
     modification-queue becoming full. This allows for this store to be unavailable for short periods of time without a
-    {@link StoreUnavailableException} being thrown, however if the store does not become available before the queue
-    fills, then a {@link StoreUnavailableException} is eventually thrown.
+    {@link StoreUnavailableException} being thrown; however if the store does not become available before the queue
+    fills, then a {@link StoreUnavailableException} is thrown.
     * @implSpec
-    * The default implementations returns a completed stage with the value {@code Boolean.TRUE}
-    * @return stage that when complete signals if the store is available
+    * The default store implementations return a completed stage with the value {@code Boolean.TRUE}.
```
Comment (oraNod) [inline core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java:223]: @wburns could you also please check this line? "default store implementations return" or should it be "default store implementation returns"? I presume the former as Infinispan provides multiple store implementations.

Comment (wburns) [reply core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java:223]: No this is referring to the default implementation of this method. TBH I think it should just be "The default implementation returns .."

Comment (oraNod) [reply core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java:223]: thanks @wburns I see that now :facepalm: was probably over-complicating things...

File: core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java]: Seems okay to me.

**Comment by wburns** [comment]
Integrated into master, thanks @oraNod !

**Pull request merged**